src/appleseed/renderer/kernel/rendering/pixelrendererbase.cpp

```diff
@@ -51,22 +54,66 @@ PixelRendererBase::PixelRendererBase()
 {
 }
 
+PixelRendererBase::PixelRendererBase(
+    const Frame&        frame,
+    const size_t        thread_index)
+  : m_invalid_pixel_count(0)
+{
+    m_invalid_sample_aov_index = frame.create_extra_aov_image("invalid_samples");
+
+    if ((thread_index == 0) && (m_invalid_sample_aov_index == size_t(~0)))
+    {
+        RENDERER_LOG_WARNING(
+            "could not create invalid samples diagnostic, maximum number of aovs (" FMT_SIZE_T ") reached.",
+            MaxAOVCount);
+    }
+}
+
 void PixelRendererBase::on_tile_begin(
     const Frame&    frame,
     Tile&           tile,
     TileStack&      aov_tiles)
 {
+    if (m_invalid_sample_aov_index != size_t(~0))
+        m_invalid_sample_diagnostic.reset(new Tile(tile.get_width(), tile.get_height(), 1, PixelFormatUInt8));
 }
 
 void PixelRendererBase::on_tile_end(
     const Frame&    frame,
     Tile&           tile,
     TileStack&      aov_tiles)
 {
+    if (m_invalid_sample_aov_index != size_t(~0))
+    {
+        const size_t width = tile.get_width();
+        const size_t height = tile.get_height();
+
+        for (size_t y = 0; y < height; ++y)
+        {
+            for (size_t x = 0; x < width; ++x)
+            {
+                Color<uint8, 1> sample_state;
+                m_invalid_sample_diagnostic->get_pixel(x, y, sample_state);
+
+                Color4f color(1.0f, 0.0f, 1.0f, 1.0f);
+
+                if (sample_state[0] == CorrectSampleHint)
+                {
+                    tile.get_pixel(x, y, color);
+                    // 20% of luminance
+                    color.rgb().set(((color.r + color.g + color.b) / 3.0f) * 0.2f);
```

Comment:
You can use `luminance(color)` which takes into account human perception of colors.